playground/backend/internal/cloud_bucket/precompiled_objects.go

```diff
@@ -0,0 +1,287 @@
+	defer cancel()
+
+	precompiledObjects := make(SdkToCategories, 0)
+	bucket := client.Bucket(BucketName)
+
+	dirs, err := cd.getPrecompiledObjectsDirs(ctx, targetSdk, bucket)
+	if err != nil {
+		return nil, err
+	}
+	for objectDir := range dirs {
+		infoPath := filepath.Join(objectDir, MetaInfoName) // helping file with information about this object
+		rc, err := bucket.Object(infoPath).NewReader(ctx)
+		if err != nil {
+			logger.Errorf("Object(%q).NewReader: %v", infoPath, err.Error())
+			continue
+		}
+		data, err := ioutil.ReadAll(rc)
+		if err != nil {
+			logger.Errorf("ioutil.ReadAll: %v", err.Error())
+			continue
+		}
+		precompiledObject := ObjectInfo{}
+		err = json.Unmarshal(data, &precompiledObject)
+		if err != nil {
+			logger.Errorf("json.Unmarshal: %v", err.Error())
+			continue
+		}
+		for _, objectCategory := range precompiledObject.Categories {
+			if targetCategory == "" || targetCategory == objectCategory { //take only requested categories
+				appendPrecompiledObject(precompiledObject, &precompiledObjects, objectDir, objectCategory)
+			}
+		}
+		rc.Close()
+	}
+	return &precompiledObjects, nil
+}
+
+// PutPrecompiledObjectsToCategory adds categories with precompiled objects to protobuf object
+func PutPrecompiledObjectsToCategory(categoryName string, precompiledObjects *PrecompiledObjects, sdkCategory *pb.Categories) {
+	category := pb.Categories_Category{
+		CategoryName:       categoryName,
+		PrecompiledObjects: make([]*pb.PrecompiledObject, 0),
+	}
+	for _, object := range *precompiledObjects {
+		category.PrecompiledObjects = append(category.PrecompiledObjects, &pb.PrecompiledObject{
+			CloudPath:   object.CloudPath,
+			Name:        object.Name,
+			Description: object.Description,
+			Type:        object.Type,
+		})
+	}
+	sdkCategory.Categories = append(sdkCategory.Categories, &category)
+}
+
+// getPrecompiledObjectsDirs finds directories with precompiled objects
+func (cd *CloudStorage) getPrecompiledObjectsDirs(ctx context.Context, targetSdk pb.Sdk, bucket *storage.BucketHandle) (map[string]bool, error) {
+	prefix := targetSdk.String()
+	if targetSdk == pb.Sdk_SDK_UNSPECIFIED {
+		prefix = ""
+	}
+	it := bucket.Objects(ctx, &storage.Query{
+		Prefix: prefix,
+	})
+	objectDirs := make(map[string]bool, 0)
+	for {
+		attrs, err := it.Next()
+		if err == iterator.Done {
+			break
+		}
+		if err != nil {
+			return nil, fmt.Errorf("Bucket(%q).Objects: %v", BucketName, err)
+		}
+		path := attrs.Name
+		if isPathToPrecompiledObjectFile(path) {
+			precompiledObjectDir := filepath.Dir(path)
+			_, ok := objectDirs[precompiledObjectDir]
+			if !ok {
+				objectDirs[precompiledObjectDir] = true
+			}
+		}
+	}
+	return objectDirs, nil
+}
+
+// appendPrecompiledObject add precompiled object to the common structure of precompiled objects
+func appendPrecompiledObject(objectInfo ObjectInfo, sdkToCategories *SdkToCategories, pathToObject string, categoryName string) {
```

Comment:
PutPrecompiledObjectsToCategory already put some object in some structure, what for this? I understand that this function puts categories into a common structure, is it right?

Reply:
It is misunderstanding because PutPrecompiledObjectsToCategory is for controller
I moved it there